build-cran-package.sh

```diff
@@ -0,0 +1,97 @@
+#!/bin/sh
```

Comment:
Run at CI and check the diff: https://github.com/microsoft/LightGBM/pull/3188#discussion_r449729842

Reply:
good news: this works (https://github.com/microsoft/LightGBM/pull/3188/checks?check_run_id=848076075)

![image](https://user-images.githubusercontent.com/7608904/86867015-ab9cdc00-c097-11ea-908b-7714ccaa0ada.png)

bad news: I realize that adding this check means we need to pin CI to a specific version of autoconf: https://github.com/microsoft/LightGBM/pull/3188/commits/012f876ab8742f369c8a5d86475015fb61c1f28c

Reply:
Ok it took a few tries, but I think I got pinning the autoconf version to work: https://github.com/microsoft/LightGBM/pull/3188/commits/9e290387879beeaded1f5e42d73ad92bf3cf9daf

Reply:
nope, that didn't work.

![image](https://user-images.githubusercontent.com/7608904/86873241-84e4a280-c0a3-11ea-9869-a588ddc4914f.png)

I'm having a hard time getting `autoconf`  to produce identical scripts on Mac and Linux, even with identical versions. I'm not sure if we can :/

Maybe some of Autoconf's other dependencies are different between the two environments.

I'm worried that this test is going to become one of those things in CI that we have to patch a lot and that blocks other PRs.

Unless you have ideas for workarounds, I think we should just generate the `configure` script on Linux and only do this check on Linux. It's most important for Linux because CRAN does not create binary packages for Linux.

Reply:
> Unless you have ideas for workarounds, I think we should just generate the configure script on Linux and only do this check on Linux. It's most important for Linux because CRAN does not create binary packages for Linux.

Ah, thank you very much for the investigation!

Of course, checking this only on Linux is enough, I believe. But maybe I'm missing something, how will it help in case when a maintainer modifies source file and re-generates config with macOS or Windows? Our check on Linux will not match their results.

Reply:
I think we will have to document how to update it on Linux. I think it would be easy enough to do in a Docker container, and the error message raised in CI can even say something like "please follow the 'regenerate configure' instructions in R-package/README.md"

Reply:
To be honest, I'm non sure it will bring any profit to us: occidental commiters will simply abandon their PRs if we ask them to use Docker with a particular environment.

I just have one idea. What if we re-generate configure at Linux CI, diff it, and in case of inconsistency [ask to update it inside Docker with Linux? and] show the updated content asking to paste it manually?

Reply:
> will simply abandon their PRs if we ask them to use Docker with a particular environment

I don't agree with this. The use of `configure.ac` +  `Autoconf` is not something that a casual contributor is likely to touch. I don't think it is too much to say "this script must be generated in a Linux environment, using Autoconf 2.69". And I don't think that it is an unfair requirement...Docker is widely used and many free public cloud options exist to get access to a small Ubuntu environment.

We can put a script like this in the repo

```shell
#!/bin/bash

# recreates 'configure' from 'configure.ac'
# this script should run on Ubuntu 18.04

AUTOCONF_VERSION=2.69-11

apt update
apt-get install \
    --no-install-recommends \
    -y \
        autoconf=${AUTOCONF_VERSION}

cp VERSION.txt R-package/
autoconf \
    --output R-package/configure \
    R-package/configure.ac
```

And then in the README document this one-line command that people can run.

```shell
docker run \
    -v $(pwd):/opt/LightGBM \
    -t ubuntu:18.04 \
    /bin/bash -c "cd /opt/LightGBM && ./R-package/recreate-configure.sh"
```

And in the case where someone says "I refuse to do this or am unable to", any of the maintainers here (including me) can update `configure` and push a commit onto the open PR.

I like that option a lot more than printing the output of `git diff` to CI logs and asking people to copy them and paste them into the script. I think that will lead to mistakes and frustration.

Reply:
I just pushed my proposed changes for this in https://github.com/microsoft/LightGBM/pull/3188/commits/c673b4fd9fabaf882933d67ac37436b56cf81725

Reply:
All your points make sense!

But I'm really afraid to scare our contributors. I remember it was not so easy to explain users that they should fix typos in https://github.com/microsoft/LightGBM/blob/master/include/LightGBM/c_api.h and then just run `python ./helpers/parameter_generator.py` but not touch https://github.com/microsoft/LightGBM/blob/master/docs/Parameters.rst directly.

> is not something that a casual contributor is likely to touch

Hm, given that we should list there all files by their full name and all compiler flags, I expect many patches from users to fix some edge cases for compiler flags in the near future and also every new added or renamed source file requires re-generating configure.

> the output of git diff to CI logs and asking people to copy them and paste them into the script. I think that will lead to mistakes and frustration.

For the sake of avoiding mistakes we can do it via git patches.

Reply:
> given that we should list there all files by their full name and all compiler flags, 

I think you are confusing the files. We're talking about people editing only `configure.ac`, which does not list any of the files by name and only has a small handful of the compiler flags that we have to search for. Are you thinking of `Makevars.in`?

Only changes to `configure.ac` would trigger the check that we're talking about in this thread.

Reply:
Oh, really? Good to hear!

Yes, I thought that filenames are embedded by this command `AC_CONFIG_FILES([src/Makevars])`. Now I see that actually it produces just the following line `ac_config_files="$ac_config_files src/Makevars"`.

Reply:
yep! `Makevars.in` is checked into the repo, and then `Makevars` is created at runtime by `configure` when the user installs the library.

Sorry if I wasn't clear explaining that before.

Given that, since we only need to regenerate `configure` from `configure.ac`, do you think the Docker solution is ok? I think it is friendly than applying a git patch. I personally don't know how to apply a git patch, and I would rather run a one-liner with Docker than push to CI, wait for the patch to be generated, then copy it and apply it to a followup commit.

Reply:
Thanks for making it clear!

I think it's a matter of taste. Someone with already installed locally and quite familiar with Docker will prefer one variant, and someone else will not waste their time installing additional software and would rather go with another variant.

However, since it is expected that such commits will be quite rarely, I believe any option is OK. Also, I expect that this updating process anyway will be done by someone from maintainers 😄 . So it doesn't matter what we will write in the README.

Reply:
ok thank you! I am going to leave it how it is then. Anyone with Docker or who can get easy access to a Linux environment another way (like with AWS, DigitalOcean, Google Cloud, Azure, etc.) can follow those directions.

Reply:
I think that `R CMD check` is running `autoreconf`. Since `configure.ac` was changed to read in the version from a file, I think that file has to be bundled in the CRAN package for `autoreconf` to work.

![image](https://user-images.githubusercontent.com/7608904/87234944-cb841680-c39b-11ea-9a2c-076fc72cf9dc.png)

I just pushed a commit that does that and that fixes the NOTE about unrecognized top-level files: https://github.com/microsoft/LightGBM/pull/3188/commits/752ad5f1efe87f550725894e3556d448951a8062

Reply:
It seems that didn't work :/

investigating...

![image](https://user-images.githubusercontent.com/7608904/87235403-66332400-c3a1-11ea-93c4-9e6e1216b629.png)

Reply:
I just simplified this a bit with the changes from #3216 https://github.com/microsoft/LightGBM/pull/3188/commits/9d5504feb168069935e209147a5d7ac474ec7abb

I also tried moving VERSION.txt  into `src/` instead of `inst/` to try to fix the problem above. Let's see if it works.

Reply:
seems that that worked!